lib/rules/media-feature-name-unit-allowed-list/index.js

```diff
@@ -0,0 +1,99 @@
+'use strict';
+
+const report = require('../../utils/report');
+const ruleMessages = require('../../utils/ruleMessages');
+const validateOptions = require('../../utils/validateOptions');
+const { TokenType } = require('@csstools/css-tokenizer');
+const { isTokenNode } = require('@csstools/css-parser-algorithms');
+const {
+	isMediaFeaturePlain,
+	isMediaFeatureRange,
+	parse,
+} = require('@csstools/media-query-list-parser');
+const validateObjectWithArrayProps = require('../../utils/validateObjectWithArrayProps');
+const { isString } = require('../../utils/validateTypes');
+const matchesStringOrRegExp = require('../../utils/matchesStringOrRegExp');
+const flattenArray = require('../../utils/flattenArray');
+const atRuleParamIndex = require('../../utils/atRuleParamIndex');
+
+const ruleName = 'media-feature-name-unit-allowed-list';
+
+const messages = ruleMessages(ruleName, {
+	rejected: (unit, name) => `Unexpected unit "${unit}" for name "${name}"`,
+});
+
+const meta = {
+	url: 'https://stylelint.io/user-guide/rules/media-feature-name-unit-allowed-list',
+};
+
+/** @type {import('stylelint').Rule} */
+const rule = (primary) => {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, {
+			actual: primary,
+			possible: [validateObjectWithArrayProps(isString)],
+		});
+
+		if (!validOptions) {
+			return;
+		}
+
+		root.walkAtRules(/^media$/i, (atRule) => {
+			const mediaQueryList = parse(atRule.params);
+
+			mediaQueryList.forEach((mediaQuery) => {
+				mediaQuery.walk((entry) => {
+					if (!isMediaFeaturePlain(entry.node) && !isMediaFeatureRange(entry.node)) {
+						return;
+					}
+
+					const name = isMediaFeatureRange(entry.node)
+						? entry.node.name.name.value[1].toString()
+						: entry.node.name.toString();
+
+					const nameKey = Object.keys(primary).find((nameIdentifier) =>
+						matchesStringOrRegExp(name, nameIdentifier),
+					);
+
+					if (!nameKey) {
+						return;
+					}
+
+					const nameList = flattenArray(primary[nameKey]);
+
+					if (!nameList) {
```

Comment:
aside: not sure how I could add a meaningful example here for code coverage

Reply:
[suggest] We can refactor the code to improve performance and readability as follows:

- Use `Object.entries()` instead of `Object.keys()`
- Memoize a return value of `Object.entries()`
- Extract a tiny internal utility
- Avoid using `flattenArray()`

```js
		if (!validOptions) {
			return;
		}

		const primaryPairs = Object.entries(primary);
		const primaryUnitList = (/** @type {string} */ featureName) => {
			for (const [name, unit] of primaryPairs) {
				if (matchesStringOrRegExp(featureName, name)) return [unit].flat();
			}

			return undefined;
		};
```

```js
		const featureName = node.name.getName();
		const unitList = primaryUnitList(featureName);

		if (!unitList) {
			return;
		}
```